src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs

```diff
@@ -53,33 +53,22 @@ public abstract partial class Workspace
 
         private protected void ClearOpenDocuments()
         {
-            List<DocumentId> docIds;
             using (_stateLock.DisposableWait())
             {
-                docIds = _projectToOpenDocumentsMap.Values.SelectMany(x => x).ToList();
-            }
```

Comment:
we would initially grab the lock (in a non-cancellable fashion) just to grab a copy of the list of IDs.  We'd then go an have to process each ID, grabbing the lock again, and mutating the workspace.   This means N+1 acquisitions of the lock.  It also opens things up for races (imo) because we now may be clearing out open-document-info while it's possible the doc to get reopened which will then update the info we then clear from thsi thread.  While presumably very unlikely, this gives me the oogies.

NOte: auditing all the code in ClearOpenDocument_NoLock (and below), the work we do is all very lightweight (just clearing out collections, and disconnecting event handlers), so it seems reasonable to hold the lock the entire time while doing this.  Fine-grained locking doesn't seem to help us here.

Reply:
@CyrusNajmabadi If I had to guess why this might have been done (and to be very clear: I don't like what I'm about to write), is clearing documents for a specific project potentially meant that an open file that was using that project as its context meant we had to go change the context to some other open project, and then we ended up calling into VS APIs and deadlocked. But I think I fixed that awhile back because it meant removing projects wasn't thread-safe and we wanted that as a part of https://github.com/dotnet/roslyn/pull/26931.

So we should be on the lookout for deadlocks here, but if we find one we should just fix it.